### PR TITLE
chore(analytics): Generalize file size analytics

### PR DESCRIPTION
**Pull request by kegluneq**
Move file size analytics code into a reusable module.

See #5312, #5314

### CODE REVIEW

**Comment by kegluneq** [comment]
Yegor volunteered to review this (thanks!), reassigning.

File: modules_dart/payload/hello_world/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    entry_points:
+        - web/index.dart
+
+- $dart2js:
+    minify: true
+    commandLineOptions:
+    - --show-package-warnings
+    - --trust-type-annotations
+    - --trust-primitives
+    - --enable-experimental-mirrors
```
Comment (yjbanov) [inline modules_dart/payload/hello_world/pubspec.yaml]: Do we need experimental mirrors?

File: modules_dart/payload/hello_world/web/index.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  // the shadowDom of this root component.
+
+  // The selector of the component passed in is used to find where to insert the
+
+  // application.
+
+  // You can use the light dom of the <hello-app> tag as temporary content (for
+
+  // example 'Loading...') before the application is ready.
```
Comment (yjbanov) [inline modules_dart/payload/hello_world/web/index.dart]: I don't think these comments are useful for the purposes of measuring code size

File: gulpfile.js [hunk 3]
```diff
@@ -1053,35 +1085,9 @@ gulp.task('!bundle.copy', function() {
...
-          path.relative(path.join('dist', 'js', 'bundle'), file.path).replace('\.gz', '');
-      analytics.bundleSize(filePath, file.contents.length, compressionLevel);
-      if (printToConsole) {
-        console.log(`  ${filePath} => ${file.contents.length} bytes (${compressionLevel})`)
-      }
-    }
-  }
+gulp.task('!bundles.js.checksize', function(done) {
+  let reportSize = require('./tools/analytics/reportsize');
+  return reportSize('dist/js/bundle/**', {printToConsole: false, reportAnalytics: true});
```
Comment (IgorMinar) [inline gulpfile.js]: we want to print to console but only at level2 just like we did before. Can you preserve that please?

File: gulpfile.js [hunk 4]
```diff
@@ -584,6 +585,37 @@ gulp.task('test.unit.dart', function(done) {
               });
 });
 
+// Dart Payload Size Test
+// This test will fail if the size of our hello_world app goes beyond one of
+// these values when compressed at the specified level.
+// Measure in bytes.
+const _DART_PAYLOAD_SIZE_LIMITS = {
+  'uncompressed': 375 * 1024,
+  'gzip level=6': 105 * 1024
```
Comment (IgorMinar) [inline gulpfile.js]: why level 6? is that what internal servers use? I in js I considered level=2 to be the main level, but I'm happy to adjust if there is evidence that our web servers default to level=6

**Comment by IgorMinar** [comment]
the rest looks fine

File: modules_dart/payload/hello_world/pubspec.yaml (hunk 1 above)
Comment (kegluneq) [inline modules_dart/payload/hello_world/pubspec.yaml]: Nope, just copied over from `modules/playground`

File: gulpfile.js (hunk 3 above)
Comment (kegluneq) [inline gulpfile.js]: Good catch, I completely missed that in the original source. Will reinstate.

File: gulpfile.js (hunk 4 above)
Comment (kegluneq) [inline gulpfile.js]: @yjbanov do you know if this is specified in our internal SLA? I figured "default" is what we would be checking against, and when I compared this value against running `gzip` on my machine `level=6` was the closest value.

**Comment by kegluneq** [comment]
Updated. Note that this *should* now fail on Travis due to https://github.com/angular/angular/issues/5419 (which Victor is fixing now)

File: gulpfile.js (hunk 4 above)
Comment (yjbanov) [inline gulpfile.js]: Apache uses zlib default, [which is level 6](http://www.zlib.net/manual.html) as of now.

Comment (yjbanov) [inline gulpfile.js]: That's also the default gzip, and yes, it level 6 is what I'm using for our SLA.

**Comment by IgorMinar** [comment]
CI unhappy because you are trying to use `let` in analytics.js which runs non-transpiled. please use `var` instead.

**Comment by kegluneq** [comment]
This PR does not update analytics.js, which already used `let`. Is there some reason reportsize.js cannot use `let` & `const` but analytics.js can?

**Comment by kegluneq** [comment]
Removing `let` & `const` statements from gulpfile.js seems to have fixed things. Travis is still upset due to dartdoc bug & failing the size check, which should be cleared up soon.

**Pull request closed without merging**